pkg/component/openvpnclient.go

```diff
@@ -1,97 +1,242 @@
 	}
 }
 
-func (o OpenVPNClient) Configure(overwrite bool) error {
-	files := o.getFileMappings()
-	if len(o.OpenVPNClient.Users) == 0 {
-		log.Fatalf("No users defined in furyagent config file passed (clusterComponent.openvpn-client.users)")
+func (o OpenVPNClient) CreateUser(clientName string) error {
+	if o.Exists(filepath.Join(OpenVPNClientPath, clientName+".crt")) {
+		return errors.New(fmt.Sprintf("client certificate for %s already exists", clientName))
 	}
-	for _, v := range o.OpenVPNClient.Users {
-		path := filepath.Join(o.OpenVPNClient.TargetDir, v)
-
-		log.Println("Creating directory for: ", v)
-		if _, err := os.Stat(path); os.IsNotExist(err) {
-			e := os.MkdirAll(path, 0755)
-			if e != nil {
-				return e
-			}
-		}
-		log.Println("Downloading CA.crt and TA.key for: ", v)
-		err := o.DownloadFilesToDirectory(files, path, OpenVPNPath, overwrite)
-		if err != nil {
-			return err
-		}
-		defer os.Remove(filepath.Join(path, OpenVPNClientCaKey)) // clean up ca.key
-		log.Println("Creating client certs for: ", v)
-		err = o.createClientCertificates(v, path)
-		if err != nil {
-			return err
-		}
+	filenames := []string{
+		OpenVPNCaCert,
+		OpenVPNCaKey,
+		OpenVPNTaKey,
+	}
+	log.Println("Downloading ca.crt, ca.key and ta.key")
+	files, err := o.DownloadFilesToMemory(filenames, OpenVPNPath)
+	if err != nil {
+		return err
+	}
+	log.Println("Creating client cert for: ", clientName)
+	clientCert, err := o.createClientCertificate(clientName, files[OpenVPNCaCert], files[OpenVPNCaKey])
+	if err != nil {
+		return err
+	}
+	type openVPNClientConfig struct {
+		Server     []string
+		CACert     string
+		ClientCert string
+		ClientKey  string
+		TLSAuthKey string
+	}
+	clientConfig := openVPNClientConfig{
+		Server:     o.OpenVPN.Servers,
+		CACert:     string(files[OpenVPNCaCert]),
+		ClientCert: string(clientCert[clientName+".crt"]),
+		ClientKey:  string(clientCert[clientName+".key"]),
+		TLSAuthKey: string(files[OpenVPNTaKey]),
+	}
+	delete(clientCert, clientName+".key")
+	log.Println("Uploading client cert for: ", clientName)
+	if err = o.UploadFilesFromMemory(clientCert, OpenVPNClientPath); err != nil {
+		return err
+	}
+	t := template.Must(template.New("openVPNClientConfig").Parse(openVPNClientConfigTmpl))
+	err = t.Execute(os.Stdout, clientConfig)
+	if err != nil {
+		return err
 	}
 	return nil
 }
 
-func (o OpenVPNClient) Init(dir string) error {
-	return nil
-}
-
-func (o OpenVPNClient) createClientCertificates(username, path string) error {
-	caCert, caKey, err := pki.TryLoadCertAndKeyFromDisk(path, "ca")
+func (o OpenVPNClient) createClientCertificate(clientName string, ca, key []byte) (map[string][]byte, error) {
+	certs, err := certutil.ParseCertsPEM(ca)
 	if err != nil {
-		return err
+		return nil, err
 	}
+	caCert := certs[0]
+	caKey, err := certutil.ParsePrivateKeyPEM(key)
 	clientCertConfig := certutil.Config{
-		CommonName:   username,
+		CommonName:   clientName,
 		Organization: []string{"SIGHUP s.r.l."},
```

Comment:
For future version/considerations... Should this value be static?

Reply:
Right now, we don't leverage the Organization field in the certificate (and I don't know if OpenVPN takes care of it all). As of now, it's mostly a kind of branding. 😄